File: lib/wal.ts
import { promises as fs } from "fs";
import { open, constants } from "fs";
import path from "path";

class WriteAheadLog {
  private readonly LOG_FILE_PATH: string;
  private byteOffset: number;

  constructor(logFileName: string) {
    this.LOG_FILE_PATH = path.resolve(logFileName);
    this.byteOffset = 0;

    // Ensure the log file exists and is ready for appending
    this.ensureLogFileExists();
  }

  public async writeInsertLog(key: string, value: string): Promise<void> {
    try {
      await this.writeLog("INSERT", key, value);
    } catch (error) {
      console.error("Error writing insert log:", error);
    }
  }

  public async writeUpdateLog(key: string, value: string): Promise<void> {
    try {
      await this.writeLog("UPDATE", key, value);
    } catch (error) {
      console.error("Error writing update log:", error);
    }
  }

  public async writeDeleteLog(key: string): Promise<void> {
    try {
      await this.writeLog("DELETE", key, "TOMBSTONE");
    } catch (error) {
      console.error("Error writing delete log:", error);
    }
  }

  public async clearLog(): Promise<void> {
    try {
      // Clear the log file by truncating it to zero bytes
      await fs.writeFile(this.LOG_FILE_PATH, "");
      this.byteOffset = 0;
    } catch (error) {
      console.error("Error clearing log:", error);
    }
  }

  public async close(): Promise<void> {
    // No explicit close needed in Node.js since `fs.promises` handles resource cleanup
    console.log("WriteAheadLog closed.");
  }

  private async writeLog(
    operation: string,
    key: string,
    value: string
  ): Promise<void> {
    const logEntry = `${this.byteOffset}, ${operation}: ${key}: ${value}\n`;
    const bytes = Buffer.from(logEntry);

    // Append the log entry to the file
    await fs.appendFile(this.LOG_FILE_PATH, bytes);

    // Update the byte offset
    this.byteOffset += bytes.length;

    // Synchronize the file to disk
    await this.fsync();
  }

  private async fsync(): Promise<void> {
    // In Node.js, there's no direct equivalent to `force(true)` in Java.
    // However, using `fs.promises.writeFile` or `fs.promises.appendFile` ensures data is flushed to disk.
    // No additional action is required here.
  }

  private async ensureLogFileExists(): Promise<void> {
    try {
      // Check if the file exists, and create it if it doesn't
      await fs.access(this.LOG_FILE_PATH, constants.F_OK);
    } catch (error) {
      // File does not exist, create it
      await fs.writeFile(this.LOG_FILE_PATH, "");
    }
  }
}

export default WriteAheadLog;

File: lib/wal.go
package main

import (
	"fmt"
	"os"
	"sync"
)

type WriteAheadLog struct {
	logFilePath string
	file        *os.File
	byteOffset  int64
	mu          sync.Mutex // To ensure thread-safe writes
}

func NewWriteAheadLog(logFileName string) (*WriteAheadLog, error) {
	file, err := os.OpenFile(logFileName, os.O_APPEND|os.O_CREATE|os.O_WRONLY|os.O_SYNC, 0644)
	if err != nil {
		return nil, fmt.Errorf("error opening log file: %w", err)
	}

	return &WriteAheadLog{
		logFilePath: logFileName,
		file:        file,
		byteOffset:  0,
	}, nil
}

func (wal *WriteAheadLog) WriteInsertLog(key, value string) {
	wal.writeLog("INSERT", key, value)
}

func (wal *WriteAheadLog) WriteUpdateLog(key, value string) {
	wal.writeLog("UPDATE", key, value)
}

func (wal *WriteAheadLog) WriteDeleteLog(key string) {
	wal.writeLog("DELETE", key, "TOMBSTONE")
}

func (wal *WriteAheadLog) ClearLog() {
	wal.mu.Lock()
	defer wal.mu.Unlock()

	// Truncate the file to clear its contents
	if err := wal.file.Truncate(0); err != nil {
		fmt.Println("Error clearing log:", err)
		return
	}

	// Reset the file offset
	if _, err := wal.file.Seek(0, 0); err != nil {
		fmt.Println("Error seeking to beginning of log file:", err)
		return
	}

	wal.byteOffset = 0
}

func (wal *WriteAheadLog) Close() error {
	wal.mu.Lock()
	defer wal.mu.Unlock()

	if err := wal.file.Close(); err != nil {
		return fmt.Errorf("error closing log file: %w", err)
	}
	return nil
}

func (wal *WriteAheadLog) writeLog(operation, key, value string) {
	wal.mu.Lock()
	defer wal.mu.Unlock()

	logEntry := fmt.Sprintf("%d, %s: %s: %s\n", wal.byteOffset, operation, key, value)
	bytesWritten, err := wal.file.WriteString(logEntry)
	if err != nil {
		fmt.Printf("Error writing %s log: %v\n", operation, err)
		return
	}

	wal.byteOffset += int64(bytesWritten)

	// Ensure data is flushed to disk
	if err := wal.fsync(); err != nil {
		fmt.Printf("Error syncing log file: %v\n", err)
	}
}

func (wal *WriteAheadLog) fsync() error {
	return wal.file.Sync()
}
